### ISSUE TITLE
Not able to compile liveCV on Mac OSX

### ISSUE THREAD

**Issue by anmoluppal** [livecv/livecv#61]
I followed the instructions, and have OpenCV and Qt already installed on my machine, However while going through instructions, the `qmake -r` didn't worked for me, So I used `qmake -r ..` and a `MakeFile` was successfully generated in the build directory, So after running `sudo make` it generates the following error code: 

```
ld: library not found for -lippicv
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [../application/liblcvlib.1.0.0.dylib] Error 1
make: *** [sub--Users-anmoluppal-Developement-testRepos-livecv-lib-lcvlib-pro-make_first] Error 2
```

**Comment by dinusv** [comment]
Hi Anmol, 

I don't have access to a Mac OS environment, so unfortunately it would be difficult for me to say what the problem might be. Based on your error, I would say there might be 2 variants:

 * You might be using Open CV version 3, instead of 2.4
 * There's a linkage problem between qmake and Open CV. Check if you can build a separate minimal qmake project linking to Open CV. I think for Mac OS you can use pkg-config to link to Open CV within your qmake file:

```
CONFIG += link_pkgconfig 
PKGCONFIG += opencv 
```

**Comment by benlau** [comment]
Hi @anmoluppal ,

How do you install opencv? If you are compiled by source, you could try to edit /usr/local/lib/pkgconfig/opencv.pc and remove the word "-lippicv" in the file.

**Comment by anmoluppal** [comment]
Yes I am using OpenCV 3.1.0 instead of 2.x, As the build requirements stated `Open CV 2.4 or higher`. 

I installed the OpenCV from source and it works fine, However after removing the `-lippicv` from `opencv.pc` I got the errors as:

```
../../../plugins/lcvcore/src/qmatdraw.cpp:108:9: error: no member named 'line' in namespace 'cv'
    cv::line(
    ~~~~^
../../../plugins/lcvcore/src/qmatdraw.cpp:147:9: error: no member named 'rectangle' in namespace 'cv'
    cv::rectangle(
    ~~~~^
../../../plugins/lcvcore/src/qmatdraw.cpp:186:9: error: no member named 'circle' in namespace 'cv'
    cv::circle(
    ~~~~^
../../../plugins/lcvcore/src/qmatdraw.cpp:235:9: error: no member named 'ellipse' in namespace 'cv'
    cv::ellipse(
    ~~~~^
../../../plugins/lcvcore/src/qmatdraw.cpp:292:17: error: no member named 'fillPoly' in namespace 'cv'
            cv::fillPoly(
            ~~~~^
../../../plugins/lcvcore/src/qmatdraw.cpp:313:17: error: no member named 'fillConvexPoly' in namespace 'cv'
            cv::fillConvexPoly(
            ~~~~^
6 errors generated.
make[2]: *** [qmatdraw.o] Error 1
make[1]: *** [sub--Users-anmoluppal-Developement-testRepos-livecv-plugins-lcvcore-make_first-ordered] Error 2
make: *** [sub--Users-anmoluppal-Developement-testRepos-livecv-plugins-make_first] Error 2
```

**Comment by dinusv** [comment]
Yes, than Open CV 3.x is the problem. There was a miscommunication on my end, I should've mentioned that Open CV 3 is not currently supported. 

'2.4 or higher versions' was mostly referred to minor version releases. They've changed the structure of the library from Open CV 3, so once you switch to version 3 it's very difficult to keep backwards compatibility with the 2nd one.

**Comment by dinusv** [comment]
I've updated the main README. Hopefully by the end of the year I will switch Live CV to Open CV 3

**Comment by anmoluppal** [comment]
So, Do I need to downgrade my Opencv environment to 2.x ? But that doesn't seems legit as I have my other projects based on OpenCV3.x

**Comment by hanhha** [comment]
I'm compiling for Mac and Opencv3 also. You can pass this by adding:
```
diff --git a/lib/include/qmat.h b/lib/include/qmat.h
index 1c6d213..2e8be75 100644
--- a/lib/include/qmat.h
+++ b/lib/include/qmat.h
@@ -19,6 +19,7 @@

 #include <QQuickItem>
 #include "opencv2/core/core.hpp"
+#include "opencv2/imgproc.hpp"
 #include "qlcvglobal.h"

 class QMatAccess;
```
However, after that I got errors relating to FeatureDectector and DescriptionExtractor redefined. After commenting out these class declarations, I got more errors relating to either cv::cv or cv::cv::cv namespace that I needed to remove a bundle of "cv::" before types.
And then I got these errors: 
```
cannot define or redeclare 'registerTypes' here because namespace 'cv' does not enclose namespace 'Lcvfeatures2dPlugin'
void Lcvfeatures2dPlugin::registerTypes(const char *uri){
     ~~~~~~~~~~~~~~~~~~~~~^
```
and
```
/Users/hha/Scratchs/livecv/plugins/lcvfeatures2d/src/lcvfeatures2d_plugin.cpp:51: error: use of undeclared identifier 'uri'
    qmlRegisterType<QKeyPoint>(                   uri, 1, 0, "KeyPoint");
                                                  ^
```
to be continued ^^ ... I'm investigating further.

**Comment by anmoluppal** [comment]
when I `make`ed with the given changes I get 

```
../../../plugins/lcvfeatures2d/src/qfeaturedetector.h:24:7: error: definition of
      type 'FeatureDetector' conflicts with typedef of the same name
class FeatureDetector;
      ^
/usr/local/include/opencv2/features2d.hpp:206:19: note: 'FeatureDetector'
      declared here
typedef Feature2D FeatureDetector;
                  ^
In file included from ../../../plugins/lcvfeatures2d/src/lcvfeatures2d_plugin.cpp:30:
../../../plugins/lcvfeatures2d/src/qdescriptorextractor.h:26:7: error: 
      definition of type 'DescriptorExtractor' conflicts with typedef of the
      same name
class DescriptorExtractor;
      ^
/usr/local/include/opencv2/features2d.hpp:213:19: note: 'DescriptorExtractor'
      declared here
typedef Feature2D DescriptorExtractor;
```

**Comment by hanhha** [comment]
Attach is the patch I made to pass. Currently only lcvfeatures2d is not able to be linked, I don't know why.
Summary in patch:
- I used QTCreator to load project and build. I've not configured qmake to be easy to call in terminal like in README yet. But I think it should be fine if we don't use QTCreator. 
- Explicitly call pkg-config to get linked libraries for OpenCV. Note that I install OpenCV and pkg-config via macport. If you didn't, just use "LIBS += -lopencv_highgui -lopencv_<all of opencv libraries>". They are same.
- OpenCV 3 doesn't support DenseFeatureDetector and GoodFeaturesToTrackDetector so that I temporarily comment out them. I would come back to implement similar features if needed.
- I can compile lcvfeatures2d plugin successfully but linking due to an unknown stupid reason so that I comment out this plugin temporarily to get all compiled and linked. 
These were what I made, I'm not sure if they work for you :D

[opencv3_macos_qtcreator.txt](https://github.com/livecv/livecv/files/453951/opencv3_macos_qtcreator.txt)

**Comment by dinusv** [comment]
Hello hanhha, thank you for your help. I've currently migrated Live CV to Open CV3 in the dev branch, I might still have some compiler fixes I need to look for, but the file you posted was quite helpful. 

I've currently disabled OpenCV contrib features, since I'm thinking of configuring them as optional.

**Comment by Azique** [comment]
Hello,

built it without any problem, application launches, but can't import modules. I get following error:
`"line 2: module "lcvcore" is not installed `
(opencv 2.4.13.2, qt 5.7)

tried to build 'dev' branch (with opencv 3), but got this:
`src/qdocumentcodeinterface.cpp:99:13: error: non-const lvalue reference to type
      'QTextCursor' cannot bind to a temporary of type 'QTextCursor'
            QTextCursor()
            ^~~~~~~~~~~~~
src/qabstractcodehandler.h:29:22: note: passing argument to parameter
      'cursorChange' here
        QTextCursor& cursorChange
                     ^
`

**Comment by dinusv** [comment]
Hi Azique,

Live CV expects the lcvcore module to be located within the <application_path>/plugins/lcvcore. There you should have the library together with a qmldir file. They are placed there automatically during the build stage, but maybe due to some reason they weren't built or they're in a different path, so you should check that first.

The lcvcore module is then dependent on Open CV libraries together with lcvlib library, so they should either be in the <application_path> or within the PATH environment variable so Live CV can find them.

The dev branch might not be currently stable accross all compilers, I'm looking to do that in the following period and add a new release. In the meantime you can try the latest commit where I addressed your issue a26ac11964c5ef6a1a510f12297bebd0c8e27e1f
